packages/amplify_datastore/android/src/main/kotlin/com/amazonaws/amplify/amplify_datastore/AmplifyDataStorePlugin.kt

```diff
@@ -15,37 +15,142 @@
+        var modelSchemas: List<Map<String, Any>>? = null
+        if (request.containsKey("modelSchemas") && request["modelSchemas"] is List<*>) {
+            modelSchemas = request["modelSchemas"].safeCastToList()
+        } else {
+            prepareError(flutterResult,
+                         java.lang.Exception(FlutterDataStoreFailureMessage.MALFORMED.toString()),
+                         FlutterDataStoreFailureMessage.CASTING.toString())
+        }
+
+        val modelProvider = FlutterModelProvider.instance
+        val flutterModelSchemaList =
+                modelSchemas!!.map { modelSchema -> FlutterModelSchema(modelSchema) }
+        flutterModelSchemaList.forEach { flutterModelSchema ->
+            modelProvider?.addModelSchema(
+                    flutterModelSchema.name,
+                    flutterModelSchema.convertToNativeModelSchema()
+            )
+        }
+        Amplify.addPlugin(AWSDataStorePlugin(modelProvider!!))
+        flutterResult.success(null)
     }
-  }
 
-  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
-    channel.setMethodCallHandler(null)
-  }
+    @VisibleForTesting
+    fun onQuery(flutterResult: Result, request: HashMap<String, Any>) {
+        var modelName = request["modelName"] as String
+        var queryOptions: QueryOptions = QueryOptionsBuilder.fromSerializedMap(request)
+        val plugin = Amplify.DataStore.getPlugin("awsDataStorePlugin") as AWSDataStorePlugin
+        plugin.query(
+                modelName,
+                queryOptions,
+                {
+                    var results: List<Map<String, Any>> =
+                            it.asSequence().toList().map { model: Model? ->
+                                FlutterSerializedModel(model as SerializedModel).toMap()
+                            }
+                    LOG.info("Number of items received " + results.size)
+                    Handler(Looper.getMainLooper()).post {
+                        flutterResult.success(results)
+                    }
+                },
+                {
+                    LOG.info("MyAmplifyApp + Query failed.$it")
+                    prepareError(flutterResult, it, FlutterDataStoreFailureMessage.QUERY.toString())
+                }
+        )
+    }
+    
+    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
+        channel.setMethodCallHandler(null)
+    }
+
+    private fun checkArguments(@NonNull args: Any): Map<String, Any> {
+        if (args !is Map<*, *>) {
+            throw java.lang.Exception("Flutter method call arguments are not a map.")
+        }
+        return args.safeCastToMap()!!
+    }
+
+    private fun prepareError(@NonNull flutterResult: Result, @NonNull error: Exception,
+                             @NonNull msg: String) {
+        LOG.error(msg, error)
+        var errorMap: HashMap<String, Any> = HashMap()
+
+        var localizedError = ""
```

Comment:
Instead of empty strings, would it be better to have some sort of default message - perhaps asking customers to check logs?

Reply:
`localizedError` and `recoverySuggestion` seems like specialized messaging, having a localized error or recovery suggestion to say that we don't know what happened seems wrong. Maybe we can say "Error not Available" or "We don't have a suggestion right now for this error"? `errorString` should still be there I believe. Thoughts?

Reply:
I like "Error not available"